ops/relation.py

```diff
@@ -0,0 +1,548 @@
+# Copyright 2021 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import logging
+import semantic_version as semver
+
+from ops.charm import CharmEvents
+from ops.framework import (
+    Object, EventSource, EventBase, StoredState
+)
+
+logger = logging.getLogger(__name__)
+
+
+class Provider(Object):
+    """Manages relations of resource providers
+
+    A :class:`Provider` object manages relations of a resource provider
+    charms, with charms that consume those resources. When a consumer
+    charm joins the relation a provider object informs it of the type
+    and version of the resource provided. Providers may also be
+    instrumented to inform consumer charms of any relavent
+    information, for example configuration settings. This is done
+    using the `provides` argument of the :class:`Provider` constructor.
+
+    Any charm that provides a resource may choose to do so through the
+    :class:`Provider` object simpliy by instantiating it in the charm's
+    `__init__` method, as follows
+
+    Example::
+
+        self.provider = Provider(self, relation_name, provides)
+        ...
+        self.provider.ready()
+
+    It is important to invoke `ready()` on the :class:`Provider`
+    object, in order to let the consumer charm know that the provider
+    is serving requests. This is done by setting a boolean flag
+    `ready` in the data forwarded to the consumer charm. A provider
+    charm may toggle this flag by invoking `unready()` when it
```

Comment:
Using ready/unready during upgrade is a little bit odd.
Specifically, 'during upgrade'. Relation data is only propagated to the other units at the completion of a successful hook. Which leaves a little bit to be desired for "if you're doing work during an upgrade, you can't really just do 'unready()' at the start and 'ready()' at the end", because you won't actually send the unready data until the hook exits (during which time you're doing the rest of the upgrade steps).

It would be plausible to have some form of interaction of "I'm going to go unready" and then wait for associated applications to ack that fact, and then use the fact that it was ack'd to actually do the steps, and then set things back to ready.

Reply:
Wouldn't `unready()` and `ready()` be useful in `on.pre_series_upgrade` and `on.post_series_upgrade` respectively ? 

I see your point with regard to `on.upgrade_charm`. For this reason your latter suggestion is interesting and I would like to explore it further. The way I understood it is set `unready()` and call `event.defer()`. Have I understood this right ? Will application data set by `unready()` get propagated after an `event.defer()`. If this understanding is not correct I would be grateful if you can elaborate.

Reply:
I understand from our last discussion (26 April 2021) we are not planning to make any changes here at the present time but leave open the possibility of revisiting this issue in the future.